conjure-go-client/httpclient/client.go

```diff
@@ -223,15 +220,10 @@ func (c *clientImpl) initializeRequestHeaders(ctx context.Context) http.Header {
 	return headers
 }
 
-func joinURIAndPath(baseURI, reqPath string) (string, error) {
-	uri, err := url.Parse(baseURI)
-	if err != nil {
-		return "", werror.Wrap(err, "failed to parse request URL")
-	}
-
+func joinURIAndPath(baseURI, reqPath string) string {
+	fullURI := strings.Trim(baseURI, "/")
```

Comment:
What's the desired contract for treatment of slashes? This will trim any prefix slashes (which probably isn't an issue since typically a protocol component is required) and also trim any trailing slashes, including if there are multiple in a row (which should probably not occur, but not sure if it's technically disallowed or just not common).

Although this is certainly an edge case, would like to have a better understanding of where `baseURI` and `reqPath` come from and what requirements/assumptions are posed on them to understand how this interacts with that behavior.

Reply:
Good question. I updated this to be `TrimRight` to be more semantically correct, although to your point I doubt there are prefix slashes anyways.

This request flow stems from the CGR defined Client (the private [clientImpl](https://github.com/palantir/conjure-go-runtime/blob/929262581a760b34308c018681c8f63466d17ee6/conjure-go-client/httpclient/client.go#L51) specifically), where the `baseURI` comes from the configured URIs of the client, either through config or using the [WithBaseURLs](https://github.com/palantir/conjure-go-runtime/blob/929262581a760b34308c018681c8f63466d17ee6/conjure-go-client/httpclient/client_params.go#L370-L376) param. The baseURLs are usually constructed to include the context-path at the end, which can also be empty.
For example, the `baseURI` could be of the following form:
```
// no context path OR empty
https://hostname:port

// single context path
https://hostname:port/api

// multi context path
https://hostname:port/foo/api
```
All of the above mutations exist as test cases as well and could exclude the `port` for a single hostname.

The `reqPath` is intended on being the extension to the `baseURI` and is included in every client request as a `RequestParam`. For conjure generated clients, this would look like the following from [conjure-go](https://github.com/palantir/conjure-go/blob/aaad4d3e22ebf4e48816fb1597fb7153dde97895/integration_test/testgenerated/client/api/services.conjure.go#L91). From my experience, it seems safe to assume this path is already encoded and should be taken as is rather than try to perform any additional encoding. 

So based on the above, I believe it should be safe to simply concatenate the two fields, while taking into account any additional path separators (`/`)

Reply:
Thanks for the extra context. So if `baseURI` and the `context-path` come from configuration, my question is around how we treat leading or trailing slashes in each case. Roughly:

1. `baseURI` = `"//foo//"`, `context-path` = `""`
2. `baseURI` = `"http://foo//`, `context-path` = `"//bar//"`
3. `baseURI` = `"http://foo/`, `context-path` = `"/bar/"`

etc.

It looks like the general contract we're going with is that when we perform joins of segments we will ensure that there's a single `/` that separates the segments, and we may trim and specified leading or trailing `/` characters to do so if needed. I think that's fine, but just wanted to be sure we're in agreement with that. It also seems like this is specifically for when we join segments -- for example, if `baseURI` is `http://foo//bar` and `context-path` is `baz`, then I assume we would produce `http://foo//bar/baz` (we'll do the join, but don't do anything about internal content).

I think this is all fine, but just want to make it explicit -- whenever we modify/normalize user-specified values, I think it's good to be explicit about how we might modify them and to think through all the cases/implications of that.

Reply:
100% agree that being explicit with our contract is ideal so we can avoid any ambiguity when we hit issues such as this one. All of your statements above are indeed correct, even the case where there is an additional segment in the baseURI + context-path -> `https://foo//bar/baz`. This just means there was a mistake on the user side and would surface as a failed request.

So in summary, the contract is essentially the following:
1. The `baseURI` must contain a scheme (`https://`, `tcp://`, etc). We do not directly handle non-scheme based urls, or at least will not mutate it.
2. The `baseURI+contextPath` should not end in a trailing `/` (even if it's `n * /`) and if it does we will strip them all.
3. If a `reqPath` exists and is not just `/`, then we will append a `/` to the `baseURI+contextPath` and append the `reqPath`.
4. The `reqPath` should begin with a single `/` if it exists and is not a single `/`

The one piece that we don't handle is trailing `/` on the reqPath, so it would be allowed to pass the following:
`baseURI+contextPath` -> `https://host/context-path`
`reqPath` -> `/foo/bar/`
`fullURI` -> `https://host/context-path/foo/bar/`

And even further:
`reqPath` -> `/foo/bar///`
`fullURI` -> `https://host/context-path/foo/bar///`

I could imagine stripping the trailing `/` from the reqPath as well, at least up to 1 trailing slash, but I am less concerned about this as well since it's typically a user error.